app/controllers/tweet_controller.rb

```diff
@@ -0,0 +1,113 @@
+class TweetController < ApplicationController
+
+  get '/tweets' do
+    if !is_logged_in?
+      redirect '/login'
+    else
+      @tweets = Tweet.all
+      erb :'/tweets/index'
+    end
+  end
+
+  get '/errors/login' do
+    erb :'/errors/login'
+  end
+
+  get '/errors/signup' do
+    erb :'/errors/signup'
+  end
+
+  get '/errors/tweets_show' do
+    erb :'/errors/tweets_show'
+  end
+
+  get '/tweets/new' do
+    if !is_logged_in?
+      redirect '/login'
+    else
+      erb :'/tweets/new'
+    end
+  end
+
+  get '/tweets/show' do
+    if !is_logged_in?
+      redirect '/login'
+    else
+      erb :'/tweets/show'
+    end
+  end
+
+  get '/tweets/:id' do
+    if !is_logged_in?
+      redirect '/login'
+    else
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'/tweets/show'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if !is_logged_in?
+      redirect '/login'
+    else
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == current_user.id
+        erb :'/tweets/edit'
+      else
+        erb :'/errors/tweets/show'
+      end
+    end
+  end
+
+  post '/tweets' do
+    if params[:content] == ""
+      redirect '/tweets/new'
+    end
+    if !is_logged_in?
+      redirect '/user/login'
+    end
+    if !!@user = User.find_by(email: params[:email])
+      redirect '/user/login'
+    end
+    @tweet = Tweet.create(content: params[:content], user_id: current_user.id)
+    redirect "/tweets/#{@tweet.id}"
+  end
+
+  patch '/tweets/:id' do
+    if params[:content] == ""
+          @tweet = Tweet.find_by_id( params[:id])
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+    if !!@user = User.find_by(email: params[:email])
+      redirect '/user/login'
+    end
+    if !is_logged_in?
+      redirect '/user/login'
+    end
+    @tweet = Tweet.find_by_id( params[:id])
+    if @tweet.user_id == current_user.id
+      @tweet.content = params[:content]
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      erb :'/errors/tweets/show'
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    if !!@user = User.find_by(email: params[:email])
+      redirect '/user/login'
+    end
+    if !is_logged_in?
+      redirect '/login'
+    else
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == current_user.id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.